[addonsinfoprovider] Addon for providing addon-info of other addons

andrewfg opened the pull request:
This addon provides an AddonInfoProvider service which produces a list of AddonInfo information for all the other addons within the openhab-addons main distribution .kar file.

In particular it provides the AddonInfo 'countries', 'connection' and 'discovery-methods' fields for all addons (even those which have not yet been installed). To give a specific example, this addon is expected to be used by the Addon Suggestion Finder service as described [here](https://github.com/openhab/openhab-core/pull/3806).

This addon scans the `addon.xml` information of ALL OH addons during the Maven build phase, and it creates an XML file which is built into this addon in the form of a resource file. When this addon is loaded, it reads the resource file and establishes an `AddonInfoProvider` service which produces the list of addons AddonInfo information from their respective addon.xml file.

Signed-off-by: Andrew Fiddian-Green <software@whitebear.ch>

andrewfg commented:
Ping @mherwege

andrewfg commented:
Note that CI build fails because it depends on changes in core https://github.com/openhab/openhab-core/pull/3806

mherwege commented:
@andrewfg Thank you, great work. I am sorry I have not been reactive lately. Work got me occupied. I hope I will be able to test both PR’s in the next few days.

One feedback already, I think you should exclude the generated addons.xml from the  PR, as it should be created at build.

andrewfg commented:
> exclude the generated addons.xml from the PR, as it should be created at build

Yes that makes sense. I need to read up on gitIgnore I suppose :(

jlaur commented on bundles/org.openhab.misc.addonsuggestionfinder/README.md:
````diff
@@ -0,0 +1,70 @@
+The following is an example for the discovery XML description for the Philips Hue bridge.
+
+```xml
+<addon:addon id="groovyscripting" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
````
```suggestion
<addon:addon id="hue" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```

andrewfg replied:
Fixed

jlaur commented on bundles/org.openhab.misc.addonsuggestionfinder/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## Addon Developer Notes
+
+If you want to your addon to scan the user's system then you need to include additional fields in your `src/main/resources/OH-INF/addon.xml` file.
```
This should be documented here, i.e. through the openhab-docs repository:
https://www.openhab.org/docs/developer/addons/addon.html#xml-structure-for-add-on-definitions

A link could be provided here, but probably the contents should be removed to avoid double maintenance.

andrewfg replied:
Agreed. But let us keep it here for now, until we make more progress on the total package.

jlaur reviewed (changes requested):
Looking forward to dig deeper into this and trying it out. Looks promising! I have added some quick feedback about the documentation.

andrewfg commented:
@mherwege I have a question to you: Currently I have the functionality split as follows..

1. The special addon produces a list of candidates
2. The core scans the candidates and produces a filtered list of actual suggestions
3. The core delivers that filtered list via a rest api call

I am thinking that I should shift the functionality 2. from the core to the special addon. There would be two benefits — the core PR becomes smaller and therefore easier for OH core maintainers to overview and approve. The specific scanning an filtering functionality can become more generic. => WDYT?

mherwege commented:
@andrewfg I think the separation between core and addons you have right now is the correct separation. If I understand it well, core will base all its suggestions on info it gets from AddonInfo. This info can come from different sources, depending on where the addons are sourced from. E.g. if the AddonInfo received for marketplace addons also contains the discovery info, nothing would stop from these being suggested as well. If an external addon source is configured, it would have a similar effect. Moving the logic to the addons repo would mean this would have to be replicated.
Also, imagine another type of suggestions: addons that have country in their AddonInfo. These could for instance be alarm services or something like that. It would be easy enough to create another type of suggestion service in the same framework in core.

mherwege commented on bundles/org.openhab.misc.addonsuggestionfinder/src/main/java/org/openhab/misc/addonsuggestionfinder/AddonSuggestionInfoProvider.java:
```diff
@@ -0,0 +1,81 @@
+        AddonInfoList addonInfoList = reader.readFromXML(xml);
+        if (addonInfoList != null) {
+            candidateAddonInfos.addAll(addonInfoList.getAddons().stream()
+                    .filter(a -> !a.getDiscoveryMethods().isEmpty()).collect(Collectors.toSet()));
```
I would not filter. Allowing all to pass through also provides the country and connection info to the AddonInfo, so we will be able to use it in the UI for filtering as well. That would solve the main issue preventing implementing https://github.com/openhab/openhab-webui/issues/2050

andrewfg replied:
Ok.

andrewfg replied:
Just thinking about this further. Currently we maintain a `Set<AddonInfo>` (both in the code in this PR and the OH-Core PR) whereby the unique filter for such set is the addon UID (the concatenation of addon type and name); however it is conceivable that some addons may be available via two addon providers (e.g. via OH main distribution and OH marketplace); and therefore two such `AddonInfo`s would be 'munged' into on entry in the set. => WDYT?

Note we do this 'munging' already in the case that an addon may have both a upnp and an mdns discovery method; whereby in this case it absolutely proper that there should be only one addon UID in the final suggestion set..

mherwege replied:
See my comment here: https://github.com/openhab/openhab-core/pull/3806#issuecomment-1780795755

mherwege replied:
The addon uid is different depending on the source of the addons. You are right for karaf provided addons. But for others, it may be something like: `marketplace:addonname`.

andrewfg replied:
Ok.

mherwege replied:
I am not sure about the best solution yet. The registry return any addon with the id it finds in the providers: https://github.com/openhab/openhab-core/blob/f2c20af94458371adecd7adc4630312165fc9fed/bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/AddonInfoRegistry.java#L71
That may have to be modified to return a "merged" version.
As described in the comment in the core PR, creating some kind of reference uid field so the suggestions can be linked to the original source could help with the merging.

mherwege commented on bundles/org.openhab.misc.addonsuggestionfinder/src/main/java/org/openhab/misc/addonsuggestionfinder/AddonSuggestionInfoProvider.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public @Nullable AddonInfo getAddonInfo(@Nullable String id, @Nullable Locale locale) {
+        return candidateAddonInfos.stream().filter(a -> a.getId().equals(id)).findFirst().orElse(null);
```
Shouldn't this be `getUID()`?
When comparing to the `XmlAddonInfoProvider` it ultimately matches on uid, not id.

mherwege replied:
Confirmed, id in AddonInfo should match uid in Addon. You therefore do not need a separate uid field in the suggestions, but the id field should be the uid from Addon.

andrewfg replied:
Yeah. I was confused by the argument name in the java doc, so I interpreted it wysiwyg, even though it didn't make real sense. I will change it.

andrewfg replied:
Done

mherwege commented:
@andrewfg I think the functionality of this addon is not just for addon discovery. It is through that this exposes all discovery info from the addon.xml files of all addons in the repository. But it also exposes all other fields currently not visible in OH before actually installing the addon. The most relevant ones are country and connection. I would therefore suggest to rename this addon and also change the name of the addon.
Also, I would argue this special addon should always be installed, independent of using the suggestion finder. At least it would allow creating filtering options in the UI on all the information available in the addon.xml files of all the addons in the repository.

andrewfg commented:
@mherwege I fully agree. I suggest the following..

1. Change its name to "Addon Info Provider" i.e. org.opehab.misc.addoninfoprovider
2. Add code to a default component in org.openhab.core to load it

=> Is that OK?

mherwege commented:
@andrewfg Your suggestion is perfectly fine.

andrewfg commented:
> Change its name to "Addon Info Provider" i.e. org.opehab.misc.addoninfoprovider

Correctio: actually changing it to "Addons Info Provider" i.e. org.opehab.misc.addonsinfoprovider (with addons with an 's') .. partly to avoid a class name clash, and partly for more clarity on its purpose.

andrewfg commented:
@mherwege addon changed name, and changed the description of this PR to match

andrewfg commented:
@mherwege this PR depends on changes in org.openhab.core.addon .. which are currently part of [#3806](https://github.com/openhab/openhab-core/pull/3806). However to make things simpler I suggest to split the core changes necessary for this PR (#15780) from those that are necessary for the other PR ([#3806](https://github.com/openhab/openhab-core/pull/3806)) into a third PR in core. => WDYT?

J-N-K commented:
@kaikreuzer I fully agree. Maybe it would be better to have that in openhab-distro, because we also need zwave and zigbee which are not part of this repository.

kaikreuzer commented:
Good point, @J-N-K.
@andrewfg I think the `AddonInfoProvider` could easily be moved to core and it could be adapted to simply read the aggregated addons.xml file where the distro puts it (Artifactory and/or kar file). Would that make sense to you?

mherwege commented:
> Good point, @J-N-K.
> @andrewfg I think the `AddonInfoProvider` could easily be moved to core and it could be adapted to simply read the aggregated addons.xml file where the distro puts it (Artifactory and/or kar file). Would that make sense to you?

I also think that makes a lot of sense. If you look at this PR, the main thing is indeed a build step that aggregates these addon.xml files. I started the experiment on this and didn't know how to do this accross repositories. Some guidance on that for @andrewfg would be appreciated.

andrewfg commented:
> Some guidance on that for @andrewfg would be appreciated.

@kaikreuzer yes please.

J-N-K commented:
For aggregating XMLs (in this case the openhab-addons BOM), I added

```
<build>
    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-antrun-plugin</artifactId>
          <version>1.8</version>
          <inherited>false</inherited>
          <executions>
            <execution>
              <id>create-bom</id>
              <goals>
                <goal>run</goal>
              </goals>
              <configuration>
                <target>
                  <copy file="${basedirRoot}/../../bundles/pom.xml" overwrite="true"
                    tofile="${basedirRoot}/../../bom/openhab-addons/pom.xml"/>
                  <!-- rewrite footer -->
                  <replaceregexp file="${basedirRoot}/../../bom/openhab-addons/pom.xml"
                    match="/modules[\s\S]*dependencies&gt;" replace="/dependencies&gt;"/>
                  <!-- rewrite header -->
                  <replaceregexp file="${basedirRoot}/../../bom/openhab-addons/pom.xml"
                    match="\S*parent[\s\S]*modules&gt;\S*" replace="header"/>
                  <replace file="{basedirRoot}/../../bom/openhab-addons/pom.xml">
                    <replacetoken>header</replacetoken>
                    <replacevalue><![CDATA[<parent>
    <groupId>org.openhab.addons.bom</groupId>
    <artifactId>org.openhab.addons.reactor.bom</artifactId>
    <version>${project.version}</version>
  </parent>

  <artifactId>org.openhab.addons.bom.openhab-addons</artifactId>
  <packaging>pom</packaging>

  <name>openHAB Add-ons :: BOM :: openHAB Add-ons</name>

  <dependencies>]]></replacevalue>
                  </replace>
                  <!-- rewrite content -->
                  <replace file="{basedirRoot}/../../bom/openhab-addons/pom.xml">
                    <replacetoken><![CDATA[<module>]]></replacetoken>
                    <replacevalue><![CDATA[<dependency>
      <groupId>org.openhab.addons.bundles</groupId>
      <artifactId>]]></replacevalue>
                  </replace>
                  <replace file="{basedirRoot}/../../bom/openhab-addons/pom.xml">
                    <replacetoken><![CDATA[</module>]]></replacetoken>
                    <replacevalue><![CDATA[</artifactId>
      <version>@dollar{project.version}</version>
    </dependency>]]></replacevalue>
                  </replace>
                  <replace file="{basedirRoot}/../../bom/openhab-addons/pom.xml">
                    <replacetoken>@dollar</replacetoken>
                    <replacevalue>$</replacevalue>
                  </replace>
                </target>
              </configuration>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
```

I think you can use something similar for aggregating discovery information from the `addon.xml`.

andrewfg commented:
> For aggregating XMLs (in this case the openhab-addons BOM), I added

@J-N-K thanks for the feedback. I think what you suggest is similar to what @mherwege suggested, and I implemented, where we use the Maven build process to create an addons.xml file at build time. However I think that @kaikreuzer is suggesting to do it with Java code at runtime. In which case see my post above, where I added some more precise questions.

J-N-K commented:
I think @kaikreuzer suggests to create the XML at build time and to read that from core instead of implementing an add-on that provides this information. Maybe the `KarafAddonService` could implement the provider for that.

mherwege commented:
> I think @kaikreuzer suggests to create the XML at build time and to read that from core

The addons.xml is already created in the maven addons build. I didn’t know how to get access to an xml generated in the addons repository build process from core. That’s why a special addon in the addons repository has been created. The better place would indeed be in core or distrib, if someone can help us understand how to get access to that xml file created during the addons build from another repository than the addons repository.

andrewfg commented:
> Maybe the KarafAddonService could implement the provider for that.

@J-N-K any thoughts about how?

andrewfg commented:
> simply read the aggregated addons.xml file where the distro puts it

@kaikreuzer I have been manually digging around inside 'openhab-addons-4.1.0-SNAPSHOT.kar' and I don't see any addons.xml in it. So is it perhaps embedded as a resource in some .jar file inside the .kar file? And if so which jar should I be looking for?

andrewfg commented:
Current solution is this..

```
/**
 * The {@link AddonsInfoProviderInstaller} component to install the special AddonsInfoProvider addon.
 *
 * @author Andrew Fiddian-Green - Initial contribution
 */
@NonNullByDefault
@Component(immediate = true, name = AddonsInfoProviderInstaller.SERVICE_NAME)
public class AddonsInfoProviderInstaller {

    public static final String SERVICE_NAME = "Installer for the AddonsInfoProvider special addon";

    private static final String KARAF_ADDONS_SERVICE_ID = "karaf";
    private static final String ADDONS_INFO_PROVIDER_UID = "misc" + Addon.ADDON_SEPARATOR + "addonsinfoprovider";

    private boolean addonInstalled;

    public AddonsInfoProviderInstaller() {
    }

    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
    public void addAddonService(AddonService addonService) {
        if (!addonInstalled && KARAF_ADDONS_SERVICE_ID.equals(addonService.getId())) {
            addonService.install(ADDONS_INFO_PROVIDER_UID);
            addonInstalled = true;
        }
    }

    public void removeAddonService(AddonService addonService) {
        if (addonInstalled && KARAF_ADDONS_SERVICE_ID.equals(addonService.getId())) {
            addonService.uninstall(ADDONS_INFO_PROVIDER_UID);
            addonInstalled = false;
        }
    }
}
```

kaikreuzer commented:
> We have a 'Russian doll' with three levels of stacking, so the questions are..

I think we can design this actually much simpler.

If the file(s) are created at build time, the result can be made available as a Maven artifact (see [here](https://github.com/openhab/openhab-distro/blob/3.0.0/features/distro-resources/pom.xml#L42-L68) for an example.

The distro build can then easily pick up these artifacts and package them within the openHAB distro, e.g. within `userdata/etc` or anywhere near there. This can be done by adding those files to a Karaf feature like it is done [here](https://github.com/openhab/openhab-distro/blob/3.0.0/features/distro/src/main/feature/feature.xml#L21). Maybe the [openhab-runtime-ui](https://github.com/openhab/openhab-distro/blob/b431198b21b0f435f62f5ebe671007dfafa7b53a/features/distro/src/main/feature/feature.xml#L4) feature can be used for this or alternatively, a new feature could be added in this file and then be packaged as a [boot feature](https://github.com/openhab/openhab-distro/blob/b431198b21b0f435f62f5ebe671007dfafa7b53a/distributions/openhab/pom.xml#L156-L173) in the distro.

andrewfg commented:
> design this actually much simpler.

Umm. @kaikreuzer that may look simple to you, but unfortunately not to me; I am a simple java programmer; and really have little clue about Karaf, Maven, or features. I shall need more hand holding please.

> The distro build can then easily pick up these artifacts and package them

I think still the issue is that the addons.xml file will be created in the openhab-addons workspace during the openhab-addons build; then the openhab-distro build needs to know the path to the openhab-addons workspace in order to import that file; and then the openhab-core needs to be able to find that file within the distro..

andrewfg commented:
> For aggregating XMLs (in this case the openhab-addons BOM), I added

@J-N-K (BTW) from the time stamps of the (supposedly overwritten) pom.xml file, and by including some xml comments in the original, that would disappear if the file were overwritten, I am pretty sure that your example code is never actually run; the pom.xml file remains identical before and after the build .

J-N-K commented:
Correct, that part is only executed if you do `mvn create-bom`. It is not run during the usual build.

andrewfg commented:
> do mvn create-bom

Actually do mvn create:bom I think..

But anyway it goes to prove that I am not the right person to try to do this work in the way that @kaikreuzer and you @J-N-K are asking for; I simply do not have the time or ability to go over the Maven, Artifact, Feature learning curve; perticularly when there is no useful documentation on those things. I shall have to put this PR into sleep mode unless someone can step in.

mherwege commented:
@andrewfg What you can already do is move the code for this AddonInfoProvider to core, whereby you read from an `addons.xml` file in `userdata/etc`.
The generation of the `addons.xml` file which is defined in the `pom.xml` of this addon needs to stay in this repository. I am not sure yet where the best place is. On top of the generation, that same build will need to make the resulting file an mvn artifact, as @kaikreuzer described.
A new PR in the openhab-distro repository needs to pick up the artifact and put it in `userdata/etc` so the AddonInfoProvider can read it at runtime.

I don't know much about karaf and mvn, but I am willing to try this in the next few weeks starting from the examples @kaikreuzer has provided. I will just need time for it, and time is scarce at the moment. So don't expect me to turn this around quickly.

In the mean time, just make the assumption there will be an `addons.xml` file the `userdata/etc` folder at runtime and start from there in the core PR.

kaikreuzer commented:
> I don't know much about karaf and mvn, but I am willing to try this in the next few weeks starting from the examples @kaikreuzer has provided. 

Thanks for volunteering here, @mherwege! I think the examples I linked should give a good starting point as it is fairly standard Maven functionality. Feel free to ping me whenever you feel you get stuck and things do not work as you would expect them to work.

> the openhab-core needs to be able to find that file within the distro.

@andrewfg That's really just classical reading a file from the filesystem, see e.g. [here](https://github.com/openhab/openhab-core/blob/cdbca4dd0ab590041492fb7464b9b5f61d296bfb/bundles/org.openhab.core.id/src/main/java/org/openhab/core/id/InstanceUUID.java#L48) how to directly access the openHAB userdata folder.

andrewfg commented:
@kaikreuzer / @mherwege two things..
1. See [this commit](https://github.com/openhab/openhab-core/pull/3865/commits/dd09a15c9d4436576c2b72379d8fd0af99946eda) where I have moved the code to openhab-core
2. Shall I (therefore) close this PR, or do you want to keep it open as a place holder?

J-N-K commented:
@mherwege https://github.com/J-N-K/openhab-addons/commit/a9a79b09bf5473bcdd08717f043f8ddfe48fbc6a But this needs to be refined a bit. IMO the XML should be more "standard" and the should be an XSD for that (which should be in core).

I'm currently preparing a commit for openhab-distro which installs the file in the correct location. Unfortunately I have to fix a build issue in openhab-addons before :-(

andrewfg commented:
> IMO the XML should be more "standard" and the should be an XSD for that (which should be in core).

@J-N-K can you be more precise about "more standard"? 

In the core PR have already modified the XSD for addon.xml to include the new discovery methods. So I am happy to add another XSD for the over wrapper containing a list of those.

andrewfg commented:
@J-N-K I am writing the XSD now. =>I will let you know so you can put the schema reference in the header.xml part of your PR.

andrewfg commented:
@J-N-K I committed the revised schemas [here](https://github.com/openhab/openhab-core/pull/3865/commits/ff2ddaa8d0facfd078354789498b15b91b3f88ef)

Please note the following..
1. I shortened the `addon-info-list` to just `addon-list` for sparseness
2. The schema for `addon-list` imports the schema for `addon`; so I am not sure if the `addon:` schema prefix is required, or it is implicit in the nested schema; i.e. one or other of the xml samples would apply. I think version 1 is fine => WDYT?

**Version 1** (without sub schema prefixes)
```
<?xml version="1.0" encoding="UTF-8"?>
<addon-list xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:addon-list="https://openhab.org/schemas/addon-list/v1.0.0"
	xsi:schemaLocation="https://openhab.org/schemas/addon-list-1.0.0.xsd">
  <addons>   
    <addon id="groovyscripting">
	<type>automation</type>
	<name>Groovy Scripting</name>
	<description>This adds a Groovy script engine.</description>
	<connection>none</connection>
    </addon>
...
  </addons>
</addon-list>
```

**Version 2** (with sub schema prefixes)
```
<?xml version="1.0" encoding="UTF-8"?>
<addon-list xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:addon="https://openhab.org/schemas/addon/v1.0.0"
	xmlns:addon-list="https://openhab.org/schemas/addon-list/v1.0.0"
	xsi:schemaLocation="https://openhab.org/schemas/addon-list-1.0.0.xsd">
  <addons>   
    <addon:addon id="groovyscripting">
	<type>automation</type>
	<name>Groovy Scripting</name>
	<description>This adds a Groovy script engine.</description>
	<connection>none</connection>
    </addon:addon>
...
  </addons>
</addon-list>
```

andrewfg commented:
@J-N-K in order for the `addons-list` schema to import the `addon` schema, I has to modify the latter so that the `addon` complex type is explicitly exported. I am pretty sure that the new version of the schema is perfectly valid xml, but apparently the `bindings.xjb` file is causing the build to fail with the new `addon` schema adaption. This looks like yet another learning curve that I need to jump over .. aargh .. so I think I will adopt a 'hammer' approach so that we can make progress .. either I shall deleted the `bindings.xjb` file, or I shall duplicate the `addon` schema inside the `addons-list` schema instead of trying to import it (which is not at all pretty, and certainly not immune from bugs if the underlying `addon` schema evolves, and the `addons-list` does not). Just so you know.

andrewfg commented:
Superseded by the following..
- https://github.com/openhab/openhab-addons/pull/15870
- https://github.com/openhab/openhab-distro/pull/1604
- https://github.com/openhab/openhab-core/pull/3865
